docs/utilities/idempotency.md

```diff
@@ -740,10 +832,37 @@ Below an example implementation of a custom persistence layer backed by a generi
     ```
 
 ???+ danger
-    Pay attention to the documentation for each - you may need to perform additional checks inside these methods to ensure the idempotency guarantees remain intact.
+Pay attention to the documentation for each - you may need to perform additional checks inside these methods to ensure
+the idempotency guarantees remain intact.
 
     For example, the `_putRecord()` method needs to throw an error if a non-expired record already exists in the data store with a matching key.
 
+## Testing your code
```

Comment:
I'm not sure showing this is super helpful. Unless this is an integration test against live resources querying the real table is probably not a great idea, nor a practical one.

I think we should mention two strategies:

1. Disabling the utility via env variable ([example](https://docs.powertools.aws.dev/lambda/typescript/latest/#environment-variables))

In this case, customers test their own code only and trust that the idempotency utility will do its job once deployed - this is what most customers will want to do since you don't usually test 3rd party modules.

2. How to mock the DynamoDB client without changing your handler code

In this case, customers might want to simulate specific failure cases. This is most likely for those customers who are wrapping arbitrary functions rather than their entire handler and want to see how their code reacts to it.

3. Using DynamoDB Local (not sure we need this)

---

I'd suggest to check the Python docs on this topic also

Reply:
Ha you are right, should have checked their section before. Will change

Reply:
Added similar tests as with Python. I did not add example where to set env variable, there are different ways how to do it with jest and other frameworks.